sdk/samples/iot/paho_iot_hub_pnp_component_sample.c

```diff
@@ -99,911 +53,859 @@ static az_span start_time_span;
-  {
-    LOG_ERROR("Could not get the device info data: error code = 0x%08x", result);
-    exit(result);
+    MQTTClient_freeMessage(&message);
+    MQTTClient_free(topic);
   }
-
-  // Send the MQTT message to the endpoint
-  mqtt_publish_message(publish_message.topic, publish_message.out_payload_span, SAMPLE_PUBLISH_QOS);
-
-  // Receive response for device info publish
-  mqtt_receive_message();
 }
 
-// Callback invoked by pnp functions each time it finds a property in the twin document
-static void sample_property_callback(
-    az_span component_name,
-    az_json_token* property_name,
-    az_json_reader property_value,
-    int32_t version,
-    void* user_context_callback)
+static void on_message_received(char* topic, int topic_len, const MQTTClient_message* message)
 {
-  (void)user_context_callback;
-  if (az_span_ptr(component_name) == NULL || az_span_size(component_name) == 0)
-  {
-    LOG("Property=%.*s arrived for Control component itself. This does not support writable "
-        "properties on it(all properties are on sub-components) ",
-        az_span_size(property_name->slice),
-        az_span_ptr(property_name->slice));
+  az_result rc;
 
-    az_result err_result;
-    if (az_failed(
-            err_result = pnp_create_reported_property_with_status(
-                publish_message.payload_span,
-                component_name,
-                property_name->slice,
-                append_json_token,
-                (void*)&property_value,
-                404,
-                version,
-                property_response_description_failed,
-                &publish_message.out_payload_span)))
-    {
-      LOG_ERROR(
-          "Could not create root component property error payload: error code = 0x%08x", err_result)
-    }
-    else
-    {
-      LOG_SUCCESS("Sending error status for root component property");
+  az_span topic_span = az_span_create((uint8_t*)topic, topic_len);
+  az_span message_span = az_span_create((uint8_t*)message->payload, message->payloadlen);
 
-      // Send error response to the updated property
-      mqtt_publish_message(
-          publish_message.topic, publish_message.out_payload_span, SAMPLE_PUBLISH_QOS);
-    }
-  }
-  else if (az_succeeded(sample_pnp_thermostat_process_property_update(
-               &client,
-               &sample_thermostat_1,
-               component_name,
-               property_name,
-               &property_value,
-               version,
-               &publish_message)))
+  az_iot_hub_client_twin_response twin_response;
+  az_iot_hub_client_method_request command_request;
+
+  // Parse the incoming message topic and check which feature it is for.
+  if (az_succeeded(
+          rc
+          = az_iot_hub_client_twin_parse_received_topic(&hub_client, topic_span, &twin_response)))
   {
-    LOG_SUCCESS("Updated property on thermostat 1");
+    LOG_SUCCESS("Client received a valid topic response:");
```

Comment:
nit:
```suggestion
    LOG_SUCCESS("Client received a valid topic response");
```
Since this is making more of a statement rather than adding detail

Reply:
you and eric are reading each other's minds...